packages/astro/src/vite-plugin-astro-server/devPipeline.ts

```diff
@@ -0,0 +1,128 @@
+import { Pipeline } from '../core/pipeline.js';
+import type { AstroConfig, AstroSettings, RouteData } from '../@types/astro';
+import type { ModuleLoader } from '../core/module-loader';
+import type { Environment } from '../core/render';
+import { createEnvironment, loadRenderer } from '../core/render/index.js';
+import { createResolve } from './resolve.js';
+import { RouteCache } from '../core/render/route-cache.js';
+import { isServerLikeOutput } from '../prerender/utils.js';
+import type { RuntimeMode, SSRManifest, SSRLoadedRenderer } from '../@types/astro';
+import type { LogOptions } from '../core/logger/core';
+import { Logger } from '../core/logger/core.js';
+import type { EndpointCallResult } from '../core/endpoint/index.js';
+import mime from 'mime';
+import { attachCookiesToResponse } from '../core/cookies/index.js';
+
+export default class DevPipeline extends Pipeline {
+	#settings: AstroSettings;
+	#loader: ModuleLoader;
+	#devLogger: Logger;
+	#currentMatchedRoute: RouteData | undefined;
+
+	constructor({
+		manifest,
+		logging,
+		settings,
+		loader,
+	}: {
+		manifest: SSRManifest;
+		logging: LogOptions;
+		settings: AstroSettings;
+		loader: ModuleLoader;
+	}) {
+		const env = DevPipeline.createDevelopmentEnvironment(manifest, settings, logging, loader);
+		super(env);
+		this.#devLogger = new Logger(logging);
+		this.#settings = settings;
+		this.#loader = loader;
+		this.setEndpointHandler(this.#handleEndpointResult);
+	}
+
+	setCurrentMatchedRoute(route: RouteData) {
+		this.#currentMatchedRoute = route;
+	}
+
+	clearRouteCache() {
+		this.env.routeCache.clearAll();
+	}
+
+	getSettings(): Readonly<AstroSettings> {
+		return this.#settings;
+	}
+
+	getConfig(): Readonly<AstroConfig> {
+		return this.#settings.config;
+	}
+
+	getModuleLoader(): Readonly<ModuleLoader> {
+		return this.#loader;
+	}
+
+	get logger(): Readonly<Logger> {
+		return this.#devLogger;
+	}
+
+	async loadRenderers() {
+		const renderers = await Promise.all(
+			this.#settings.renderers.map((r) => loadRenderer(r, this.#loader))
+		);
+		this.env.renderers = renderers.filter(Boolean) as SSRLoadedRenderer[];
+	}
```

Comment:
As you can notice, the side effects stay inside the pipeline, so we can better control them.